packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts

```diff
@@ -0,0 +1,110 @@
+import iam = require('@aws-cdk/aws-iam');
+import cdk = require('@aws-cdk/core');
+import { CfnCloudFrontOriginAccessIdentity } from './cloudfront.generated';
+
+export interface CloudFrontOriginAccessIdentityProps {
+  /**
+   * Any comments you want to include about the origin access identity.
+   *
+   * @default "Allows CloudFront to reach the bucket"
+   */
+  readonly comment?: string;
+}
+
+export interface ICloudFrontOriginAccessIdentity extends cdk.IResource, iam.IGrantable  {
+  /**
+   * The Origin Access Identity Name
+   */
+  readonly name: string;
+}
+
+export abstract class BaseCloudFrontOriginAccessIdentity extends cdk.Resource {
+  /**
+   * The Origin Access Identity Name (physical id)
+   */
+  public abstract readonly name: string;
+  /**
+   * Derived principal value for bucket access
+   */
+  public abstract readonly grantPrincipal: iam.IPrincipal;
+
+  /**
+   * The Arn to include in S3 bucket policy to allow CloudFront access
+   */
+  protected arn(): string {
+    return cdk.Stack.of(this).formatArn(
+      {
+        service: "iam",
+        region: "", // global
+        account: "cloudfront",
+        resource: "user",
+        resourceName: `CloudFront Origin Access Identity ${this.name}`
+      }
+    );
+  }
+}
+
+/**
+ * An origin access identity is a special CloudFront user that you can
+ * associate with Amazon S3 origins, so that you can secure all or just some of
+ * your Amazon S3 content.
+ */
+export class CloudFrontOriginAccessIdentity extends BaseCloudFrontOriginAccessIdentity implements ICloudFrontOriginAccessIdentity {
+  public static fromName(
+    scope: cdk.Construct,
+    id: string,
+    name: string): BaseCloudFrontOriginAccessIdentity {
+
+    class Import extends BaseCloudFrontOriginAccessIdentity {
+      public readonly name = name;
+      public readonly grantPrincipal = new iam.ArnPrincipal(this.arn());
+    }
+
+    return new Import(scope, id);
+  }
+
+  /**
+   * The Amazon S3 canonical user ID for the origin access identity, used when
+   * giving the origin access identity read permission to an object in Amazon
+   * S3.
+   *
+   * @attribute
+   */
+  public readonly cloudFrontOriginAccessIdentityS3CanonicalUserId: string;
+
+  /**
+   * Derived principal value for bucket access
+   */
+  public readonly grantPrincipal: iam.IPrincipal;
+
+  /**
+   * The Origin Access Identity Name (physical id)
+   */
+  public readonly name: string;
+
+  /**
+   * CDK L1 resource
+   */
+  private readonly resource: CfnCloudFrontOriginAccessIdentity;
+
+  constructor(scope: cdk.Construct, id: string, props?: CloudFrontOriginAccessIdentityProps) {
+    super(scope, id);
+
+    this.resource = new CfnCloudFrontOriginAccessIdentity(this, "Resource", {
+      cloudFrontOriginAccessIdentityConfig: {
+        comment: (props && props.comment) || "Allows CloudFront to reach to the bucket"
+      }
+    });
+    // physical id - OAI name
+    this.name = this.resource.ref.toString();
```

Comment:
Please see other examples on how physical names should be handled. It's a bit convoluted but required in order to support things like cross-account access.

Reply:
I did not find another example of a physical id which can't be specified in properties on create and is handled differently. All the vpc-id, subnet-id resources use the same method of not specify a `name` attribute in properties and excluding `"props-physical-name:` in awslint. It would be great if you could point me to a better example.